### ISSUE TITLE
Drafts with No Changes

### ISSUE THREAD

**Issue by krispharper** [release-drafter/release-drafter#871]
We just installed this and it's working great on some of our repos, but on others, after merging a PR, we get a draft release that just says, "No Changes."

![image](https://user-images.githubusercontent.com/1270239/119197074-350d9080-ba55-11eb-9b7d-95acd2b0356b.png)

Our config in `.github` looks like this

```
change-template: '* $TITLE (#$NUMBER)'
template: |
  $CHANGES
```

Do we have something misconfigured?

**Comment by krispharper** [comment]
Actually just noticed that it appears it's always missing the latest PR in the changes. I.e., if we merge one PR it says "No changes." If we then merge another PR, it will list the first PR, but not the second, etc.

**Comment by Manuel-Naix** [comment]
That is the same behaviour I see in my project. The latest PR merge triggers the action to newly draft the release but the PR merged is not considered and misses in the draft. Only the next merge ensures that the PR appears in the release draft but then again, the most recent misses.

**Comment by JimNero009** [comment]
Same for some of our projects -- not all repos, just some of them, have exactly this kind of 'n-1' behaviour.

**Comment by krispharper** [comment]
Yes, also noticed that it doesn't affect every repo. I'd say maybe half. But I haven't noticed a pattern as to which ones are affected.

**Comment by c-emil** [comment]
I noticed the same issue when I pushed a commit straight to the master branch without a PR. It updated the draft release that contained all the PRs that were merged previously with "No Changes" body.

**Comment by jetersen** [comment]
I have never experienced the n-1 maybe it is because your using the github app?

Have you tried switching to the github action?

**Comment by c-emil** [comment]
@jetersen 
> thats the intended effect. Release drafter is only able to properly categorize pull requests thanks to labels

I think we misunderstood each other. The problem is that it wipes out the body of the draft release completely and adds "No changes" in the body if I push a commit straight to the master branch. What I'd expect is that either it doesn't change anything in the draft release since it only adds changes from the PR commits, or it adds those commits in the "Other changes" (a preferable solution which was also suggested in the issues you referenced). None of this happens.

**Comment by estahn** [comment]
I'm facing a similar issue. Would be great if there is no draft created if there are no changes. I'm using `publish: true` and it started to create empty releases.

As for @c-emil issue, the content of `release-drafter` could be placed in tags similar to `<!-- release-drafter:being !--> ... <!-- release-drafter:end !-->`, which would aid in replacing content partially.

**Comment by ksolie** [comment]
@estahn Did you ever figure out a way to not create a draft if there are no changes?

**Comment by lmagyar** [comment]
FYI: In our experience, the root cause can be, that PR's associated with commits are coming from GH's GraphQL API, and that needs some time to process the same event, that triggered the release drafter itself. We added a 15s delay before running the release drafter step, ugly but works. Later we put the release drafter as the last step of our CI, that means at least 1 minute delay.
```
      - name: Wait for 15 seconds to ensure GraphQL consistency
        shell: bash
        run: sleep 15s
```

**Comment by cchanche** [comment]
Please see recent comments on #493 . Might be a dupe

**Comment by cchanche** [comment]
To clarify : this is high-effort fix with low-effort workaround.

Issue comes from GitHub's Graphql Commit-to-PR relation `associatedPullRequests`, which take some time for github to compute. When you push to `main` and trigger a release-drafter run, you sometimes end up being faster than Github's backend and may miss this relation.

To prevent this, simply delay your release-drafter run as @lmagyar suggested.

On release-drafter's side, this will definitely need to be worked-on some-day, especially if we are to support individual commits inside the release changes. Such commits will be miss-categorized as being _without a pull request_, and you'll end-up with merge commits inside the _individual commit changes_.

On a side-note, I'd love to find another way of finding about pull-requests and commits inside the remote. Using github's grapqhl is a major vendor lock-in, and prevents us from supporting (any) other remote.

EDIT : see https://github.com/release-drafter/release-drafter/pull/1408#issuecomment-4082227574 for potential implem plan

**Comment by cchanche** [comment]
Yet another instance of this issue in this very repo, on latest `master` (_to-be `v7.3.0`_ with #1570 included)...

Logs of [action run](https://github.com/release-drafter/release-drafter/actions/runs/25130007623/attempts/1) triggered when #1570 was squash-merged in commit f188d08e9e71d8903f02ca1c5e7aea645a815537 in branch master: 

```
Current runner version: '2.334.0'
Runner Image Provisioner
Operating System
Runner Image
GITHUB_TOKEN Permissions
Secret source: Actions
Prepare workflow directory
Prepare all required actions
Getting action download info
Download action repository 'release-drafter/release-drafter@master' (SHA:f188d08e9e71d8903f02ca1c5e7aea645a815537)
Complete job name: update_release_draft

with:
    config-name: release-drafter.yml
    token: ***
Parsing inputs and configuration...
Config fetched on remote "release-drafter/release-drafter@refs/heads/master".
Fetching releases from GitHub...
Found 61 releases
Draft release:
  tag_name:  v7.2.2
  name:      v7.2.2
Last release:
  tag_name:  v7.2.1
  name:      v7.2.1
Finding commits between refs/tags/v7.2.1 and refs/heads/master...
Found 2 commits.
Found 1 pull requests associated with those commits. 1 of those are merged and come from release-drafter/release-drafter : #1597
Building release payload and body...
Version increment: patch
Resolving version info based on:
   - last release: v7.2.1
   - version input: none
   - version key increment: patch
Coerce and parse versions from last release...
Parsed version from last release: 7.2.1.
Coerce and parse versions from input...
Parsed version from input: none.
Release payload built successfully
  name:                        v7.2.2
  tag:                         v7.2.2
  body:                        253 characters long
  targetCommitish:             refs/heads/master
  prerelease:                  false
  make_latest:                 true
  draft:                       true
  RESOLVED_VERSION:            7.2.2
  RESOLVED_VERSION_MAJOR:      7
  RESOLVED_VERSION_MINOR:      2
  RESOLVED_VERSION_PATCH:      2
  RESOLVED_VERSION_PRERELEASE: 
Updating existing release...
Release updated!
Set action outputs...
Outputs set!
```

The important part being : 

```
Found 2 commits.
Found 1 pull requests associated with those commits. 1 of those are merged and come from release-drafter/release-drafter : #1597
```

...should have been :

```
Found 2 commits.
Found 2 pull requests ...
```

Resulting draft release : 

<img width="943" height="353" alt="Image" src="https://github.com/user-attachments/assets/bdafc860-5af8-409b-90ff-cd86aa5cc5a5" />

Re-run of the same action ([run number 2](https://github.com/release-drafter/release-drafter/actions/runs/25130007623)) a few minutes later has the correct logs :

```
Finding commits between refs/tags/v7.2.1 and refs/heads/master...
Found 2 commits.
Found 2 pull requests associated with those commits. 2 of those are merged and come from release-drafter/release-drafter : #1597, #1570
```


... and release now looks ok (7.3.0 instead of 7.2.2 because it ended up finding the new feature):

<img width="938" height="595" alt="Image" src="https://github.com/user-attachments/assets/de7f792f-9128-44d0-868a-d899fe438465" />

cc @jetersen

**Comment by lmagyar** [comment]
"It is a general issue around GitHub and their caching. It is annoying, and I've also noticed delays not helping." https://github.com/frenck/python-tailscale/issues/1412#issuecomment-4275917823

**Comment by jetersen** [comment]
@cchanche maybe if it literally is always the last PR merged. What if we simply always fetch the last 3-5 merged PRs via rest API always and use that as reference to fix the last associated PR caching bug 🤔

**Comment by jetersen** [comment]
https://github.com/release-drafter/release-drafter/pull/1604 confirms the lag and testing confirms safety net can catch it by calling merged PRs even via graphql, so we are not solely relying on the associated PRs discovery.
